### PR TITLE
CSHARP-2861: Having incorrect credentials along with minPoolSize > 0 makes the driver to endlessly create new connections..

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
EG:   https://evergreen.mongodb.com/version/5dee7f693e8e8629004929d4

**NOTE**: Sounds like there is no way to have unclosed sockets in case if opening connection fails. I've added a test to check it: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/92/files#diff-462741f8dce658fd58f4f6dd2699ab02R36

The methods where sockets are disposed:
 - `TcpStreamFactory`: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs#L169
 - `SslStreamFactory`: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Connections/SslStreamFactory.cs#L77

**NOTE2**: If I understood the java implementation correctly, java has a little bit different implementation of this logic: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs#L255

In case if `EnsureMinSizeAsync` attempt fails, `java` doesn't start a new attempt immediately, only after 1 min delay, in the `c#` we'are trying to make new attempts right after failing. So, it means that for the case when we provide wrong credentials, we will keep trying to create new connections all the time.

I've implemented the possible fix for it and related test: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/92/files#diff-6217848c52f25851ca783e0d424c8160R267
Not 100% that it's required, we can discuss it.

**UPDATE:**
**NOTE3**:  The previous implementation didn't close a socket when authentication has failed. The **NOTE1** above tells that we will always close sockets during opening connections, but authentication error will be triggered in the next step after opening. So, in that case, opening by itself is successful, but the authentication attempt failed.

### CODE REVIEW

**Review by rstam — COMMENTED** [review]
Looks good so far.

Re-request my review once you have a final version.

**Comment by DmitryLukyanov** [comment]
This ticket is not in code review now. I will come back to it on Monday. I guess I will temporarily remove reviewers.

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs [hunk 1]
```diff
@@ -322,7 +321,16 @@ private async Task EnsureMinSizeAsync(CancellationToken cancellationToken)
...
                     // the whole point of having a min pool size is to have
                     // them available and ready...
-                    await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+                    try
+                    {
+                        await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (MongoAuthenticationException)
+                    {
+                        // if connection fails with MongoAuthenticationException we should close a socket
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs]: As I understand, this is a similar change as we have here in `GetChannel`: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Servers/Server.cs#L152 when we clear connection pool only for `MongoAuthenticationException`.

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs [hunk 2]
```diff
@@ -322,7 +321,16 @@ private async Task EnsureMinSizeAsync(CancellationToken cancellationToken)
                     // when adding in a connection, we need to open it because 
                     // the whole point of having a min pool size is to have
                     // them available and ready...
-                    await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+                    try
+                    {
+                        await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (MongoAuthenticationException)
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs]: I think it should dispose on any exception type, not just an auth exception.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs]: Not sure that we can have any other errors(except MongoAuthenticationException) which will give unclosed sockets in this line, but I think there is no harm to add this suggestion. Done

**Review by DmitryLukyanov — COMMENTED** [review]
EG: https://evergreen.mongodb.com/version/5deec2d03e8e8629004fb9a9

File: src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs (hunk 2 above)
Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs]: Any one of the handshake commands (ismaster, buildinfo, auth) could just fail for some reason (socket error, ok: 0), so I think it's the right thing to do.

**Review by rstam — APPROVED** [review]
LGTM

**Pull request closed without merging**